[git-webkit] Add clone command

JonWBedard opened the pull request:
#### 1463644b7ce9725be5e1ec33d0870244409f02cb
<pre>
[git-webkit] Add clone command
<a href="https://bugs.webkit.org/show_bug.cgi?id=254599">https://bugs.webkit.org/show_bug.cgi?id=254599</a>
rdar://107320897

Reviewed by Dewei Zhu.

Add a command to quickly clone a radar from a commit or Bugzilla.

* Tools/Scripts/libraries/webkitbugspy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/__init__.py: Import mock milestones.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/data.py: Add mock milestones.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/radar.py:
(AppleDirectoryQuery.member_dsid_list_for_group_name): Add.
(RadarModel.Milestone): Add mock Milestone object.
(RadarModel.Category): Add mock Category object.
(RadarModel.Event): Add mock Event object.
(RadarModel.Tentpole): Add mock Tentpole object.
(RadarModel.MilestoneAssociations): Add mock MilestoneAssociations object.
(RadarModel.commit_changes): Commit milestone, category, event and tentpole.
(RadarModel.milestone_associations): List MilestoneAssociations object from Milestone object.
(RadarClient.milestones_for_component): List all mock milestones.
(RadarClient.clone_radar): Mock implementation of Radarclient&apos;s clone_radar.
(Radar.__init__):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/radar.py:
(Tracker.create): Handle empty project.
(Tracker.clone): Clone radar with specified reason.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/radar_unittest.py:
* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/__init__.py:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/clone.py: Added.
(Clone.parser): User must specify problem, reason and optionally milestone
to clone radar into.
(Clone.main): Resolve the provided argument to a radar and then clone the
specified radar. Attempt to copy as much of the original radar as possible
after moving the cloned radar into the specified milestone.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/clone_unittest.py: Added.
(TestClone):

Canonical link: <a href="https://commits.webkit.org/265475@main">https://commits.webkit.org/265475@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b80f88966e8dd7df881f2029efbcf1bd62bb630b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10992 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11204 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11524 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/12639 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/10496 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/11007 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13581 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/11186 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/13421 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/11151 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/12045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/9265 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/13044 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [  ~~🧪 webkitpy~~](https://ews-build.webkit.org/#/builders/6/builds/11094 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/9338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/9935 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/13044 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/10409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/10089 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/13044 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/10524 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/11186 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [  ~~🧪 services~~](https://ews-build.webkit.org/#/builders/28/builds/10893 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/9696 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/13965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1232 "Failed to push commit to Webkit repository") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/10378 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rjepstein commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/clone.py:
```diff
@@ -0,0 +1,166 @@
+    def main(cls, args, repository, **kwargs):
+        rdar = None
+        for tracker in Tracker._trackers:
+            if isinstance(tracker, radar.Tracker):
```
is this a typo? `rdar` vs `radar`

JonWBedard replied:
No, might need to change some variable names, `radar` would conflict with `from webkitbugspy import radar`

dewei-zhu reviewed (approved):
r=me

JonWBedard committed:
[git-webkit] Add clone command
https://bugs.webkit.org/show_bug.cgi?id=254599
rdar://107320897

Reviewed by Dewei Zhu.

Add a command to quickly clone a radar from a commit or Bugzilla.

* Tools/Scripts/libraries/webkitbugspy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/__init__.py: Import mock milestones.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/data.py: Add mock milestones.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/radar.py:
(AppleDirectoryQuery.member_dsid_list_for_group_name): Add.
(RadarModel.Milestone): Add mock Milestone object.
(RadarModel.Category): Add mock Category object.
(RadarModel.Event): Add mock Event object.
(RadarModel.Tentpole): Add mock Tentpole object.
(RadarModel.MilestoneAssociations): Add mock MilestoneAssociations object.
(RadarModel.commit_changes): Commit milestone, category, event and tentpole.
(RadarModel.milestone_associations): List MilestoneAssociations object from Milestone object.
(RadarClient.milestones_for_component): List all mock milestones.
(RadarClient.clone_radar): Mock implementation of Radarclient's clone_radar.
(Radar.__init__):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/radar.py:
(Tracker.create): Handle empty project.
(Tracker.clone): Clone radar with specified reason.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/radar_unittest.py:
* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/__init__.py:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/clone.py: Added.
(Clone.parser): User must specify problem, reason and optionally milestone
to clone radar into.
(Clone.main): Resolve the provided argument to a radar and then clone the
specified radar. Attempt to copy as much of the original radar as possible
after moving the cloned radar into the specified milestone.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/clone_unittest.py: Added.
(TestClone):

Canonical link: https://commits.webkit.org/265475@main
